File: tree/671-second-minimum-node-in-a-binary-tree.cpp
// Solution 1: recursion
// N = number of nodes
// pattern: post-order
// O(N) in time (when the second minimum node
// is in the right most of the tree, then
// need to visit all the node to find it)
// O(N) in space in worst case ("skewed" tree, of
// course each node still has 2 sub nodes)
// O(h) in average, h = height of tree.
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        return helper(root, root->val);
    }

    int helper(TreeNode* node, const int & smallest) {
        if (node == NULL) return -1;
        if (node->val > smallest) return node->val;

        int geLeft = helper(node->left, smallest);
        int geRight = helper(node->right, smallest);
        if (geLeft == -1) return geRight;
        if (geRight == -1) return geLeft;
        return min(geLeft, geRight);
    }
};

// Solution 2: iterative solution of solution 1
// Please refer to Python file.

// Solution 3: another iterative solution
class Solution {
public:
    int findSecondMinimumValue(TreeNode* root) {
        if (root == NULL) return -1;
        int min_ = root->val;
        stack<TreeNode*> stack({root});
        int res = INT_MAX;
        while(!stack.empty()) {
            TreeNode* node = stack.top();
            stack.pop();
            if (min_ < node->val && node->val < res)
                res = node->val;
            if (node->left != NULL) {
                stack.push(node->left);
                stack.push(node->right);
            }
        }
        return (res == INT_MAX) ? -1 : res;
    }
};

File: practices/leetcode/tree/671-second-minimum-node-in-a-binary-tree.py
# Solution 1: iterative
# Implementation of solution 2 in C++ file.
# O(N) in space and time
class Solution:
    def findSecondMinimumValue(self, root):
        if root is None: return -1
        stack = [root]
        visited = {None: -1}
        while stack:
            node = stack[-1]

            if node.val > root.val:
                visited[node] = node.val
                stack.pop()
            elif (not node.left or node.left in visited) and \
                 (not node.right or node.right in visited):
                if visited[node.left] == -1:
                    visited[node] = visited[node.right]
                elif visited[node.right] == -1:
                    visited[node] = visited[node.left]
                else:
                    visited[node] = min(visited[node.left], visited[node.right])
                stack.pop()
            else:
                if node.left: stack.append(node.left)
                if node.right: stack.append(node.right)
        return visited[root]
